File: app/src/main/java/com/explorify/xplore/xplore_demo/RegisterActivity.java
package com.explorify.xplore.xplore_demo;

import android.app.Activity;
import android.app.DatePickerDialog;
import android.app.Dialog;
import android.app.DialogFragment;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.Exclude;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;
import com.google.firebase.database.ServerValue;
import com.google.firebase.database.ValueEventListener;
import com.squareup.picasso.Picasso;

import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

import jp.wasabeef.picasso.transformations.RoundedCornersTransformation;

/**
 * Created by nikao on 3/11/2017.
 */

public class RegisterActivity extends Activity {

    private int bYear, bMonth, bDay;
    public static FirebaseUser newUser;
    public static DatePickerDialog.OnDateSetListener myDateSetListener;

    private final int LIMIT_AGE = 16;

    private EditText fnameTxt, lnameTxt, emailTxt, telnumTxt;
    private TextView birthDate;
    private ImageView profileImg;
    private Button doneBtn;
    private Long tempTimeStamp;
    private UploadUser tempUser;

    DatabaseReference DBref = FirebaseDatabase.getInstance().getReference();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.register_layout);

        setupDateSetListener();
        getServerEpoch();

        //Init
        fnameTxt = (EditText) findViewById(R.id.reg_fname_text);
        lnameTxt = (EditText) findViewById(R.id.reg_lname_text);
        emailTxt = (EditText) findViewById(R.id.reg_email_text);
        telnumTxt = (EditText) findViewById(R.id.reg_telnum_text);
        doneBtn = (Button) findViewById(R.id.reg_done_button);

        profileImg = (ImageView) findViewById(R.id.reg_profile_image);
        //TODO CHOOSE GALLERY IMAGE OR TAKE PHOTO
        //TODO AI check for face in photo?

        birthDate = (TextView) findViewById(R.id.reg_birthdate_text);
        birthDate.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                DialogFragment newFragment = new DatePickerFragment();
                newFragment.show(getFragmentManager(), "datePicker");
            }
        });

        //Loading user data
        tempUser = createTempUser();

        //Loading data into views
        fnameTxt.setText(tempUser.getFname());
        lnameTxt.setText(tempUser.getLname());
        emailTxt.setText(tempUser.getEmail());
        Picasso.with(RegisterActivity.this)
                .load(tempUser.getProfile_picture_url())
                .transform(new RoundedCornersTransformation(
                        getResources().getInteger(R.integer.pic_big_angle),
                        getResources().getInteger(R.integer.pic_big_margin)))
                .into(profileImg);

        doneBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(checkFields()) {
                    tempUser.setBirth_date(General.GetDateLong(bYear,bMonth,bDay));
                    tempUser.setTel_num(telnumTxt.getText().toString());
                    //TODO upload image to db if changed
                    addUserEntryToDataBase(tempUser);
                }
            }
        });
    }

    private UploadUser createTempUser()
    {
        String photoUrl = " ";
        if(newUser.getPhotoUrl() != null)
            photoUrl = newUser.getPhotoUrl().toString();

        String fullName = newUser.getDisplayName();

        String[] name = {fullName,"."};
        if(fullName.contains(" "))
            name = fullName.split(" ", 2);

        return new UploadUser(newUser.getUid(), name[0], name[1], photoUrl, newUser.getEmail());
    }

    private void addUserEntryToDataBase(UploadUser user)
    {
        Map<String, Object> userData = user.toMap();

        Map<String, Object> childUpdates = new HashMap<>();
        childUpdates.put("/users/" + user.getUid(), userData);
        Log.println(Log.INFO, "BREJK", "uid: " + user.getUid());
        DBref.updateChildren(childUpdates);
        setResult(RESULT_OK);
        finish();
    }

    private class UploadUser extends User
    {
        String uid;

        public UploadUser(String uid, String fname, String lname, String profile_picture_url, String email) {
            this.uid = uid;
            this.fname = fname;
            this.lname = lname;
            this.profile_picture_url = profile_picture_url;
            this.age = 0;
            this.tel_num = " ";
            this.email = email;
            //this.tel_num = tel_num; //TODO add tel_num field that the user fills
            this.reputation = 0; //new user starts with 0 rep
        }

        public String getUid() {
            return uid;
        }

        @Exclude
        public Map<String, Object> toMap()
        {
            HashMap<String, Object> result = new HashMap<>();
            result.put("fname", this.fname);
            result.put("lname", this.lname);
            result.put("profile_picture_url", this.profile_picture_url);
            result.put("birth_date", this.birth_date);
            result.put("tel_num", this.tel_num);
            result.put("reputation", this.reputation);
            result.put("email", this.email);
            return result;
        }
    }

    private void setupDateSetListener() {
        myDateSetListener = new DatePickerDialog.OnDateSetListener() {
            @Override
            public void onDateSet(DatePicker datePicker, int year, int month, int day) {
                if (General.isNetConnected(RegisterActivity.this)) {
                    Calendar cal = Calendar.getInstance();
                    cal.setTime(new Date(tempTimeStamp));
                    int nowYear = cal.get(Calendar.YEAR);
                    //int nowMonth = cal.get(Calendar.MONTH);
                    //int nowDay = cal.get(Calendar.DAY_OF_MONTH);

                    if ((nowYear - year) >= LIMIT_AGE) {
                        bYear = year;
                        bMonth = month + 1;
                        bDay = day;

                        birthDate.setText(bYear+"/"+bMonth+"/"+bDay);
                    } else
                        Toast.makeText(RegisterActivity.this, "You must be of age " + LIMIT_AGE + " to use Xplore", Toast.LENGTH_SHORT).show(); //TODO string resources
                } else
                    General.createNetErrorDialog(RegisterActivity.this);
            }
        };
    }

    private void getServerEpoch()
    {
        DatabaseReference ref = FirebaseDatabase.getInstance().getReference();
        Map<String, Object> dateValue = new HashMap<>();
        dateValue.put("timestamp", ServerValue.TIMESTAMP);
        ref.child("date").setValue(dateValue);
        Query query = ref.child("date").child("timestamp");
        query.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                tempTimeStamp = dataSnapshot.getValue(Long.class);
            }
            @Override
            public void onCancelled(DatabaseError databaseError) {}
        });
    }

    private boolean checkFields()
    {
        //TODO check fields
        return true;
    }
}

File: app/src/main/java/com/explorify/xplore/xplore/RegisterActivity.kt
package com.explorify.xplore.xplore

import android.app.Activity
import android.app.DatePickerDialog
import android.os.Bundle
import android.util.Log
import android.widget.DatePicker
import android.widget.EditText
import android.widget.Toast

import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.Exclude
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ServerValue
import com.google.firebase.database.ValueEventListener
import com.squareup.picasso.Picasso

import java.util.Calendar
import java.util.Date
import java.util.HashMap

import jp.wasabeef.picasso.transformations.RoundedCornersTransformation
import kotlinx.android.synthetic.main.register_layout.*

/**
 * Created by nikao on 3/11/2017.
 */

class RegisterActivity : Activity(), DatePickerDialog.OnDateSetListener {

    private val ageRestriction: Int = 16

    private var bYear: Int = 0
    private var bMonth: Int = 0

    private var bDay: Int = 0
    private var tempTimeStamp: Long = 0

    init {
        getServerEpoch()
    }

    internal val DBref = FirebaseDatabase.getInstance().reference

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.register_layout)

        //Getting User info from SignInActivity
        val userId = intent.getStringExtra("userId")
        val userFullName = intent.getStringExtra("fullName")
        val userEmail = intent.getStringExtra("email")
        val userProfilePicUrl = intent.getStringExtra("photoUrl")

        //Loading data into views
        fnameEditText.setText(separateFullName(userFullName, 0))
        lnameEditText.setText(separateFullName(userFullName, 1))
        emailEditText.setText(userEmail)

        //Birth date selector
        bdateTextView.setOnClickListener {
            //Creating new DialogFragment
            val fragment = com.explorify.xplore.xplore.DatePickerDialogFragment(this, tempTimeStamp, ageRestriction)
            fragment.show(fragmentManager, "datePicker")
        }
        //TODO CHOOSE GALLERY IMAGE OR TAKE PHOTO
        //TODO AI check for face in photo?
        if(userProfilePicUrl != "") {
            Picasso.with(this@RegisterActivity)
                    .load(userProfilePicUrl)
                    .transform(RoundedCornersTransformation(
                            resources.getInteger(R.integer.pic_big_angle),
                            resources.getInteger(R.integer.pic_big_margin)))
                    .into(profileImageView)
        }

        doneButton.setOnClickListener {
            if (checkFields()) {
                //TODO upload image to db if changed
                addUserEntryToDataBase(
                        UploadUser(userId, fnameEditText.str(), lnameEditText.str(),
                                numEditText.str(), userEmail,
                                General.getDateLong(bYear, bMonth, bDay), userProfilePicUrl)
                )
            }
        }
    }

    override fun onDateSet(datePicker: DatePicker, year: Int, month: Int, day: Int) {
        if (General.isNetConnected(this@RegisterActivity)) {
            val cal = Calendar.getInstance()
            cal.time = Date(tempTimeStamp!!)
            val nowYear = cal.get(Calendar.YEAR)
            //int nowMonth = cal.get(Calendar.MONTH);
            //int nowDay = cal.get(Calendar.DAY_OF_MONTH);

            //Checking if age is OK

            if (nowYear - year >= ageRestriction) {
                bYear = year
                bMonth = month
                bDay = day

                bdateTextView.text = "$bYear/$bMonth/$bDay"
            } else
                Toast.makeText(this@RegisterActivity,
                        "You must be of age $ageRestriction to use Xplore",
                        Toast.LENGTH_SHORT).show() //TODO string resources
        } else
            General.createNetErrorDialog(this@RegisterActivity)
    }

    private fun addUserEntryToDataBase(user: UploadUser) {
        val userData = user.toMap()

        val childUpdates = HashMap<String, Any>()
        childUpdates.put("/users/" + user.uid, userData)
        Log.println(Log.INFO, "BREJK", "uid: " + user.uid)
        DBref.updateChildren(childUpdates)
        setResult(Activity.RESULT_OK)
        finish()
    }

    //TODO change init after converting User class to kotlin
    private inner class UploadUser(val uid: String, fname: String, lname: String, tel_num: String,
                                   email: String, birth_date: Int,
                                   profile_picture_url: String? = "") : User() {

        init {
            this.id = uid
            this.fname = fname
            this.lname = lname
            this.tel_num = tel_num;
            this.email = email
            this.birth_date = birth_date

            //Setting profile picture URL
            if (profile_picture_url != null) {
                this.profile_picture_url = profile_picture_url
            }

            //Every new user starts with 0 reputation
            this.reputation = 0
        }

        @Exclude
        fun toMap(): Map<String, Any> {
            val result = HashMap<String, Any>()
            result.put("fname", this.fname)
            result.put("lname", this.lname)
            result.put("profile_picture_url", this.profile_picture_url)
            result.put("birth_date", this.birth_date)
            result.put("tel_num", this.tel_num)
            result.put("reputation", this.reputation)
            result.put("email", this.email)
            return result
        }
    }

     fun getServerEpoch() {
        val ref = FirebaseDatabase.getInstance().reference
        val dateValue = HashMap<String, Any>()
        dateValue.put("timestamp", ServerValue.TIMESTAMP)
        ref.child("date").setValue(dateValue)
        val query = ref.child("date").child("timestamp")
        query.addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                tempTimeStamp = dataSnapshot.getValue(Long::class.java)
            }
            override fun onCancelled(databaseError: DatabaseError) { }
        })
    }

    private fun checkFields(): Boolean {
        //TODO check fields
        return true
    }

    //Splits a full name and returns the i-th part of it. Returns fullName back if it has no space
    private fun separateFullName(fullName: String, i: Int): String {
        var name = arrayOf<String>(fullName,"")
        if (fullName.contains(" "))
            name = fullName.split(" ".toRegex(), 2).toTypedArray()

        return name[i]
        }

    fun EditText.str() = this.text.toString() //TODO take this to general
}
